### PR TITLE
Improve decomposition for constant_pad_nd

### PR DESCRIPTION
**Pull request by isuruf**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #123661
* #137315

Replaces the previous decomposition that used a fill and a copy
using _unsafe_masked_index and removes the lowering

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @amjames @desertfire @chauhang

### CODE REVIEW

**Comment by isuruf** [comment]
The test failures are because `_unsafe_masked_index` is not in `torch._refs`. Should I move the implementations of _unsafe_masked_index and co to `torch/refs/__init__.py`?

File: torch/_decomp/decompositions.py [hunk 1]
```diff
@@ -3754,6 +3755,53 @@ def _unsafe_masked_index_put_accumulate(x, mask, indices, values):
...
+        ).reshape(-1, *[1] * (dim - 1 - i))
+        for i in range(dim)
+    ]
+
+    indices: List[Any] = [None] * input.dim()
+    for i in range(dim):
+        indices[i + nc_dim] = out_indices[i]
+
+    if builtins.all(pad <= 0 for pad in pad):
+        result = aten._unsafe_index(input, indices)
```
Comment (peterbell10) [inline torch/_decomp/decompositions.py:3788]: Could this fast path be done as a series of slices followed by a clone instead? That might allow for some better graph level optimizations which would be neat.

File: torch/_decomp/decompositions.py [hunk 2]
```diff
@@ -3754,6 +3755,53 @@ def _unsafe_masked_index_put_accumulate(x, mask, indices, values):
...
+            -pad_left[i], inp_shape[i] + pad_right[i], device=input.device
+        ).reshape(-1, *[1] * (dim - 1 - i))
+        for i in range(dim)
+    ]
+
+    indices: List[Any] = [None] * input.dim()
+    for i in range(dim):
+        indices[i + nc_dim] = out_indices[i]
+
+    if builtins.all(pad <= 0 for pad in pad):
```
Comment (peterbell10) [inline torch/_decomp/decompositions.py:3787]: For optimizations like this, you should use `torch.fx.experimental.symbolic_shapes.statically_known_true` which avoids introducing a guard.

Comment (lezcano) [reply torch/_decomp/decompositions.py:3801]: also, but [`for pad in pad`](https://en.wikipedia.org/wiki/Russell%27s_paradox) :P

Comment (isuruf) [reply torch/_decomp/decompositions.py:3801]: Is `statically_known_true` okay to have in `torch._decomp`?

Comment (peterbell10) [reply torch/_decomp/decompositions.py]: Yes, it's fine to use inside decompositions.

File: torch/_inductor/lowering.py [hunk 3]
```diff
@@ -3402,56 +3401,6 @@ def loader(idx):
     )
 
 
-@register_lowering(aten.constant_pad_nd, type_promotion_kind=None)
-def constant_pad_nd(x, padding, fill_value=0):
-    assert (len(padding) % 2) == 0
-    if all(p == 0 for p in padding):
-        return clone(x)
-
```
Comment (eellison) [inline torch/_inductor/lowering.py:3410]: This old code code is suspiciously similar to pointwise_cat lowering.. wonder if we could unify constant_pad_nd and aten.cat implementations

Comment (isuruf) [reply torch/_inductor/lowering.py:3638]: I thought making `pointwise_cat` a inductor primitive and make it a have a decomposition using `_unsafe_masked_index` just like `constant_pad_nd`, but AFAIK there's not way to call a decomposition from the lowering of `aten.cat`.

**Comment by isuruf** [comment]
The test failures are a bug in gcc 9/10. If I use gcc 8 or 11+, this test passes.

A minimal reproducer is
```
(gcc) jenkins@b7e5a22d102e:~/workspace$ cat asd.cpp                                                                                                                
#include <cassert>
#include <cstdint>
#include <cstdlib>
#include <iostream>

void kernel(const int32_t* in_ptr0,
                       int32_t* out_ptr0)
{
    {
        #pragma GCC ivdep
        for(long x0=static_cast<long>(0L); x0<static_cast<long>(3L); x0+=static_cast<long>(1L))
        {
            #pragma omp simd simdlen(8) 
            for(long x1=static_cast<long>(0L); x1<static_cast<long>(7L); x1+=static_cast<long>(1L))
            {
                auto tmp0 = (-2L) + x1;
                int64_t tmp1 = (tmp0);
                auto tmp2 = static_cast<int64_t>(0);
                auto tmp3 = tmp1 >= tmp2;
                auto tmp4 = static_cast<int64_t>(3);
                auto tmp5 = tmp1 < tmp4;
                auto tmp6 = tmp3 && tmp5;
                auto tmp7 = [&]
                {
                    auto tmp8 = in_ptr0[static_cast<long>((3L*x0) + ((((-2L) + x1) < 0L) ? (1L + x1) : ((-2L) + x1)))];
                    return tmp8;
                }
                ;
                auto tmp9 = tmp6 ? tmp7() : static_cast<decltype(tmp7())>(0);
                out_ptr0[static_cast<long>(x1 + (7L*x0))] = tmp9;
            }
        }
    }
}
int main() {
  int32_t in[20];
  int32_t out[21];
  for(int i = 0; i <20; i++) {
     in[i] = i;
  }
  kernel(in, out);
  assert(out[9] == 3);
}
```
and compiling with
```
g++ asd.cpp -O3 -march=native -fopenmp
```
fails the assert and if you remove `-fopenmp`, it passes because `simdlen(8)` directive is skipped.

**Comment by bhack** [comment]
Any update on this?

File: torch/_inductor/decomposition.py [hunk 4]
```diff
@@ -948,3 +949,17 @@ def max_pool2d_with_indices(
         padding,
     )
     return vals, indices
+
+
+# TODO: remove this when gcc 10 support is dropped in inductor
```
Comment (amjames) [inline torch/_inductor/decomposition.py:954]: Can you open an issue / link to an existing one for this? The TODO reads like there is a compiler bug in gcc 9/10, but the comment below reads like there is a bug in our codegen.

File: torch/_refs/__init__.py [hunk 5]
```diff
@@ -2850,7 +2850,8 @@ def conj(input: TensorLikeType) -> TensorLikeType:
 
 
 # This replicates at::constant_pad_nd, defined in ATen/native/PadNd.cpp
-@register_decomposition(aten.constant_pad_nd)
+# Not registering here as there's a better decomposition using
+# torch._decomp.decompositions using the private method _unsafe_masked_index.
```
Comment (amjames) [inline torch/_refs/__init__.py:2854]: nit/typo:  
```suggestion
# Not registering here as there's a better decomposition in
# torch._decomp.decompositions using the private method _unsafe_masked_index.
```

**Review by amjames — COMMENTED** [review]
Couple of minor things, but I think this is ready for a serious look.

**Comment by eellison** [comment]
clearing review queue, please re-request when ready

**Comment by isuruf** [comment]
Ready for review now

**Review by eellison — COMMENTED** [review]
look good but perf results ?

**Comment by isuruf** [comment]
With the other PR in the stack, this gives the exact same triton kernel as the lowering.

**Comment by isuruf** [comment]
@eellison any comments on this?

**Comment by eellison** [comment]
Hmm, @shunting314 recently made a change to constant pad nd lowering. these two prs might not be compatible. https://github.com/pytorch/pytorch/pull/140249. cc @shunting314 , take a look ?

**Comment by shunting314** [comment]
Decompose constant _pad_nd rather than lowering it make it hard (if possible) to do optimization in https://github.com/pytorch/pytorch/pull/140249 . With the lowering rule, Inductor is able to leverage the layout it picked for the input tensor and do inplace padding if eligible. That can save about 4 ms for gpt2 training. I don't see how this can be done if we decomp the op and use aten._unsafe_masked_index.

Should we keep this op for lowering?

**Comment by eellison** [comment]
clearing review queue but let me know how you want to follow up. potentially we could land this to aten decompositions but exclude from inductor for now.

**Comment by pianpwk** [comment]
Hi @isuruf, I was wondering if you were planning to continue work on this? It's become relevant to some work in dynamic shapes/data-dependent errors for export. If not, I'd be happy to pick it up, at least for the aten decomp side (inductor is less relevant to us).

Thanks!

**Comment by isuruf** [comment]
@pianpwk do you want `constant_pad_nd` to decomopse into `_unsafe_masked_index`? Or something else?

**Comment by pianpwk** [comment]
> @pianpwk do you want `constant_pad_nd` to decomopse into `_unsafe_masked_index`? Or something else?

Mainly it'd be nice if pad didn't guard on the pad values being < 0 or >= 0 (except the case where all <= 0, that's fine). For export the decomposition currently specializes on the guard values inferred from sample inputs, so if we export with pad=(4, -4) we can't generalize to (4, 4) at runtime.

It seems like your implication is ok with this? I might be wrong.
